### ISSUE TITLE
Bootstrap shows unexpected warning: unable to check if origin/master is old due to error

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#131296]
I am getting this on each `x.py` invocation now, since very recently:
```
warning: unable to check if origin/master is old due to error: No such file or directory (os error 2)
warning: origin/master is used to determine if files have been modified
warning: if it is not updated, this may cause files to be needlessly reformatted
```
This is in my main checkout, *not* in a worktree, so I am not quite sure what could even be so unusual about my setup that causes an error here.

Cc @rust-lang/bootstrap

**Comment by RalfJung** [comment]
A bisect points at
```
cea707d96033d5e14d3c2b287b0bedd2847db680 is the first bad commit
commit cea707d96033d5e14d3c2b287b0bedd2847db680
Author: binarycat <binarycat@envs.net>
Date:   Tue Aug 27 16:31:40 2024 -0400

    emit old upstream warning no matter the build step

 src/bootstrap/src/core/build_steps/format.rs | 4 +---
 src/bootstrap/src/core/sanity.rs             | 6 ++++++
 2 files changed, 7 insertions(+), 3 deletions(-)
```
That's https://github.com/rust-lang/rust/pull/129584.

Cc @lolbinarycat

**Comment by RalfJung** [comment]
It seems to try to read `.git/refs/remotes/origin/master` which indeed is not a file that exists here.
```
$ tree .git/refs/
.git/refs/
├── bisect
├── heads
│   ├── immediate-offset-sanity-check
│   ├── master
│   └── master.2
├── remotes
│   └── origin
│       └── HEAD
└── tags
```
I also noticed this line in `.git/packed-refs`
```
f559d6188828b738ce7e7c2e4d99bf03111336d6 refs/remotes/origin/master
```
So maybe checking these files is just futile since git doesn't always use a file to track these refs?

**Comment by RalfJung** [comment]
I think what happened is that `git gc` has been run, and that indeed cleans up most of the files from `.git/refs/remotes`.

So the current approach in bootstrap of looking at these files is not reliable. IMO we should instead look at the commit date of the most recent commit in that branch. That also avoids having to directly mess with git's internal files.

**Comment by lolbinarycat** [comment]
yeah, i'm gonna accept defeat on this one.  `git log` is kinda slow on my device, but only if the relevant files aren't in the i/o cache, and building rust takes a good while anyways.

**Comment by onur-ozkan** [comment]
https://github.com/rust-lang/rust/issues/129528 is not the case anymore since [refactor merge base logic and fix x fmt PR](https://github.com/rust-lang/rust/pull/130161). I am planning to revert https://github.com/rust-lang/rust/pull/130121 and https://github.com/rust-lang/rust/pull/129584 as they are not really required.

**Comment by onur-ozkan** [comment]
> https://github.com/rust-lang/rust/issues/129528 is not the case anymore since https://github.com/rust-lang/rust/pull/130161.

Seems like it will still be the case if upstream was configured and too old. But that could be fixed in a much better way (e.g., we could ignore upstream commit and use merge commit in current branch) compare to https://github.com/rust-lang/rust/pull/130121 and https://github.com/rust-lang/rust/pull/129584.
